### PR TITLE
Performance 5 (modules)

### PR DESCRIPTION
**Pull request by frankmayer**
Performance (and a bit of cleanup) Batch #5

The changes in this batch are all in files under `modules`.
This PR modifies code to be a bit more performant and also does some cleanup.

I have mostly done work on low hanging fruit. There are still other ways of improving, but whose involve more deep research in the code and probably more drastic changes in order to be implemented.

### CODE REVIEW

**Comment by Bakual** [comment]
This would need a lot of testing since you're sometimes now are checking variables stricter than before.
A lot of the changes also look very pointless to me, like changing `" "` to `' '` or changing `is_null` to a comparision `=== null`.
Personally I'm not in favor of doing such changes because while they might have a very tiny effect on performance, in practice you will not notice them at all. Even after 1000 instances of them.

On the other hand the chances that some bug slips in is higher than the tiny gain we get.

As long as we don't have a codestyle rule regarding those cases (which I would not be in favor of), I wouldn't merge this.

**Comment by rdeutz** [comment]
@frankmayer there are a lot of changes and you put a lot of work into it, this is much appreciated, but I have to agree with @Bakual the chance to break something somewhere is much bigger then the performance advantage we can get. If we would have proper unit tests to make such refactoring safe I would merge it but we don't so I also wouldn't merge it. 

There is a lot of work of this kind to be done in Joomla4 and I suggest to contact @mbabker or @wilsonge if you can help them.

**Comment by mbabker** [comment]
> If we would have proper unit tests to make such refactoring safe I would merge it but we don't so I also wouldn't merge it.

Then we can't ever merge anything like this outside the libraries folder unfortunately.  Modules aren't even (practically) testable because of their procedural nature; you basically include the entry file and hope for the best.

Ya it's a lot of stuff to go through, but whether it be the maintenance team on the 3.x branch or the few of us really focusing on the 4.0 branch someone's gotta do it at some point.  And admittedly things are going to be much better tested on the 3.x branch until 4.0's somewhere around a beta.

**Comment by rdeutz** [comment]
In 4 we have to do a lot of cleaning up the house stuff, I am saying spend time there it makes much more sense. If it is $var == 'alias' or $var === 'alias" doesn't makes a difference

**Comment by frankmayer** [comment]
@Bakual @rdeutz 
I understand your points.
As for the changes made, they are more or less based on best practices.
For example:

- It's better to use single quotes than double quotes if not embedding some variable within the string .
- is_null() is a function call and thus more expensive than just comparing to `null`
- The order of conditions in if statements also makes a big difference, if the result of the condition can be computed at the beginning with cheaper instructions, than the ones where multiple functions may be called.

As for the strict typing, I hand picked and checked those one by one. At points where it was unclear, I didn't change them. So most of the times strict typing was enforced when there really were strings to compare to or where a function involved in the comparison would return a string or definitely an integer.
So, a lot of work has been done in those PR's with which anyone having to deal with strict typing in future, will not have to deal with.

As for the micro-optimizations and if they are noticeable or even necessary, that is a big discussion. My personal opinion is that whatever causes a machine to do less work, is acceptable.

**Comment by frankmayer** [comment]
@rdeutz actually it does make a difference, unless you don't want strict typing implemented, which is also a valid way to look at things. However IMHO strict typing makes things a lot clearer and easier to understand. Loose typing is always a bit of unclear and error prone.

**Comment by Bakual** [comment]
@frankmayer I don't disagree with the best practices. For new code they may even make sense to a degree. But going back and refactor existing code for such microoptimisation is a waste of time of the ones writing them, the reviewers, testers, mergers and finally the other PR writer who need to solve the conflicts this PRs are going to generate.

> As for the micro-optimizations and if they are noticeable or even necessary, that is a big discussion. My personal opinion is that whatever causes a machine to do less work, is acceptable.

That's where we disagree. I can support microoptimisations for code that is run hundreds of times in a request cycle. Eg stuff related to build and parse URLs in Joomla or library stuff like JRegistry calls.
But here we are talking about modules, they are run once per cycle and most of the time even are cached. It's a waste of human ressources here.
Same if we're going to talk about administration area, there performance doesn't matter a lot as well.

So it really depends on what you are touching if it is worth or not.

**Comment by Bakual** [comment]
> is_null() is a function call and thus more expensive than just comparing to null

Btw, I read some comments that in PHP7 `is_null` is actually faster than `=== null`. Haven't tested it (not worth my time :smile)

**Comment by frankmayer** [comment]
@Bakual I can understand your human resources point micro-optimizations but not for the best practices. For best practices such changes are building blocks for future updates using more compliant code (which is less error prone, etc...).
As for where best practices or even optimizations are done, it does matter. For code compliance inside a project, all parts should be on the same level with best practices or use optimized code.
According to Joomla's stats [here](https://docs.joomla.org/Evaluators) Joomla runs on at least tenths of thousands installations. So, once you're in that ballpark, those micro-optimizations do matter when thinking in terms of wasted energy because of non-optimal code. But that's just my opinion.

BTW, those PR didn't cost me but a few hours. I just wanted a change of scenery and thought of giving back to the project. A bit of cleaning up and optimizing stuff was my idea to do that. Some super-optimizations might be an OCD of mine, but since I was already at it, I just did those, too.

If you think this PR is not worth the time, then by all means throw it away.

As for the `is_null()` being faster, that's a really nice development. Things keep changing... :smile:

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 933ab8d3bfde48b2ebc7303962783f924f58ab97<br /><br />on code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12229">issues.joomla.org/tracker/joomla-cms/12229</a>.</sub>

**Comment by frankmayer** [comment]
Sorry. Messed this one up. Will try to find a solution...

File: modules/mod_articles_categories/helper.php [hunk 1]
```diff
@@ -35,11 +35,12 @@ public static function getList(&$params)
 		$categories = JCategories::getInstance('Content', $options);
 		$category   = $categories->get($params->get('parent', 'root'));
 
-		if ($category != null)
+		if ($category !== null)
 		{
 			$items = $category->getChildren();
 
-			if ($params->get('count', 0) > 0 && count($items) > $params->get('count', 0))
+			$count = $params->get('count', 0);
```
Comment (andrepereiradasilva) [inline modules/mod_articles_categories/helper.php]: cs: please add a empty line before the if statement
also
`$items = array_slice($items, 0, $count);` (inside the if)?

File: modules/mod_feed/tmpl/default.php [hunk 2]
```diff
@@ -19,38 +19,39 @@
 {
 	$lang      = JFactory::getLanguage();
 	$myrtl     = $params->get('rssrtl');
-	$direction = " ";
+	$direction = ' ';
 
-	if ($lang->isRtl() && $myrtl == 0)
+	$isRtl = $lang->isRtl();
```
Comment (andrepereiradasilva) [inline modules/mod_feed/tmpl/default.php]: cs: please add a empty line before the if statement

File: modules/mod_feed/tmpl/default.php [hunk 3]
```diff
@@ -88,9 +89,9 @@
 		<ul class="newsfeed<?php echo $params->get('moduleclass_sfx'); ?>">
 		<?php for ($i = 0, $max = min(count($feed), $params->get('rssitems', 5)); $i < $max; $i++) { ?>
 			<?php
-				$uri   = (!empty($feed[$i]->uri) || !is_null($feed[$i]->uri)) ? trim($feed[$i]->uri) : trim($feed[$i]->guid);
-				$uri   = substr($uri, 0, 4) != 'http' ? $params->get('rsslink') : $uri;
-				$text  = !empty($feed[$i]->content) ||  !is_null($feed[$i]->content) ? trim($feed[$i]->content) : trim($feed[$i]->description);
+				$uri   = (!empty($feed[$i]->uri) || $feed[$i]->uri !== null) ? trim($feed[$i]->uri) : trim($feed[$i]->guid);
+				$uri = strpos($uri, 'http') !== 0 ? $params->get('rsslink') : $uri;
```
Comment (andrepereiradasilva) [inline modules/mod_feed/tmpl/default.php]: cs: please ident the equal sign

File: modules/mod_menu/helper.php [hunk 4]
```diff
@@ -139,9 +139,9 @@ public static function getList(&$params)
 
 				if (isset($items[$lastitem]))
 				{
-					$items[$lastitem]->deeper     = (($start?$start:1) > $items[$lastitem]->level);
-					$items[$lastitem]->shallower  = (($start?$start:1) < $items[$lastitem]->level);
-					$items[$lastitem]->level_diff = ($items[$lastitem]->level - ($start?$start:1));
+					$items[$lastitem]->deeper     = (($start?:1) > $items[$lastitem]->level);
+					$items[$lastitem]->shallower  = (($start?:1) < $items[$lastitem]->level);
+					$items[$lastitem]->level_diff = ($items[$lastitem]->level - ($start?:1));
```
Comment (andrepereiradasilva) [inline modules/mod_menu/helper.php]: cs cna we replace the `$start?:1` for `$start ?: 1`?

File: modules/mod_random_image/helper.php [hunk 5]
```diff
@@ -152,8 +149,7 @@ public static function getFolder(&$params)
 			$folder = str_replace(JPATH_BASE, '', $folder);
 		}
 
-		$folder = str_replace('\\', DIRECTORY_SEPARATOR, $folder);
-		$folder = str_replace('/', DIRECTORY_SEPARATOR, $folder);
+		$folder = str_replace(array('\\', '/'), DIRECTORY_SEPARATOR, $folder);
 
 		return $folder;
```
Comment (andrepereiradasilva) [inline modules/mod_random_image/helper.php]: just `return str_replace(array('\\', '/'), DIRECTORY_SEPARATOR, $folder);`?

**Review by andrepereiradasilva — COMMENTED** [review]
some optimizations + cs

File: modules/mod_articles_categories/helper.php (hunk 1 above)
Comment (frankmayer) [inline modules/mod_articles_categories/helper.php]: Nice catch :+1:

File: modules/mod_feed/tmpl/default.php (hunk 2 above)
Comment (frankmayer) [inline modules/mod_feed/tmpl/default.php]: OK

File: modules/mod_menu/helper.php (hunk 4 above)
Comment (frankmayer) [inline modules/mod_menu/helper.php]: OK

File: modules/mod_random_image/helper.php (hunk 5 above)
Comment (frankmayer) [inline modules/mod_random_image/helper.php]: OK

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 72e49795a38ca1b5ffcd090b6c1c48d4d1462034<br /><br />code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12229">issues.joomla.org/tracker/joomla-cms/12229</a>.</sub>

**Comment by frankmayer** [comment]
Conflicts resolved...

**Pull request merged**